src/vendor/vendor.c

```diff
@@ -125,23 +106,237 @@ int RDK_SetGroup(int group_id, kv_vector_t *params, unsigned *param_types, int *
 int DiscoverDM_ForAllComponents(char *objs_filename, char *params_filename);
-int DiscoverDM_ForComponent(rdk_component_t *rdkc, kv_vector_t *rdk_objects, kv_vector_t *rdk_params);
-void Add_NameToDM(rdk_component_t *rdkc, char *instantiated_path, char *write_status, kv_vector_t *rdk_objects, kv_vector_t *rdk_params);
-void Add_ObjectToDM(rdk_component_t *rdkc, char *schema_path, char *write_status, kv_vector_t *rdk_objects);
-void Add_ParamToDM(rdk_component_t *rdkc, char *instantiated_path, char *schema_path, char *write_status, kv_vector_t *rdk_params);
+int Discover_AllDM(kv_vector_t *rdk_objects, kv_vector_t *rdk_params);
+void Add_NameToDM(char *instantiated_path, char *write_status, kv_vector_t *rdk_objects, kv_vector_t *rdk_params);
+void Add_ObjectToDM(char *schema_path, char *write_status, kv_vector_t *rdk_objects);
+void Add_ParamToDM(char *instantiated_path, char *schema_path, char *write_status, kv_vector_t *rdk_params);
 void ConvertInstantiatedToSchemaPath(char *src, char *dest, int len);
-void AddMissingObjs(kv_vector_t *rdk_objects, kv_vector_t *rdk_params, kv_vector_t *missing_objs);
 int WriteDMConfig(char *filename, char *mode, kv_vector_t *kvv, char *comment);
-void *rdk_malloc(size_t size);
-void rdk_free(void *ptr);
 
 #ifdef    INCLUDE_LCM_DATAMODEL
 #include "lcm_rbus_datamodel.c"
 #endif
 
+#define DEV_IP_DIA "Device.IP.Diagnostics.IPPing."
+
+// Array of valid input arguments
+static char *pingtest_input_args[] =
+{
+    "Host",
+};
+
+// Array of valid output arguments
+static char *pingtest_output_args[] =
+{
+    "Status",
+    "SuccessCount",
+    "FailureCount",
+};
+
+// input results of pingtest
+typedef struct
+{
+    int request_instance;
+}pingtest_input_cond_t;
+
+// Output results of pingtest
+typedef struct
+{
+    char Status[256];
+    char SuccessCount[256];
+    char FailureCount[256];
+} pingtest_output_res_t;
+
+/*********************************************************************//**
+**
+** PingThreadMain
+**
+** Called to log the status, SuccessCount and Failure Count of Device.IP.Diagnostics.IPPing() operation
+**
+** \param   params - pointer to input condition
+**
+** \return  USP_ERR_OK if successful
+**
+**************************************************************************/
+void *PingThreadMain(void* params)
+{
+    rbusValue_t values = NULL;
+    int rbus_err, count=4;
+    pingtest_input_cond_t *cond = (pingtest_input_cond_t *) params;
+    pingtest_output_res_t results;
+    pingtest_output_res_t *res = &results;
+    kv_vector_t *output_args;
+    int err = USP_ERR_OK;
+    char *err_msg = NULL, path[256];
+
+    memset(&results, 0, sizeof(results));
+
+    // Exit if unable to signal that this operation is active
+    err = USP_SIGNAL_OperationStatus(cond->request_instance, "Active");
+    if (err != USP_ERR_OK)
+    {
+        USP_SNPRINTF(err_msg, sizeof(err_msg), "%s: USP_SIGNAL_OperationStatus() failed", __FUNCTION__);
+        goto exit;
+    }
+
+    do {
```

Comment:
Why should we poll for a value of `Device.IP.Diagnostics.IPPing.DiagnosticsState` for every 3 secs?
We could subscribe for a value change and wait on Mutex; if the new value received in the event is `Complete` then we can signal to move to the next step.